File: count-inversions/count-inversions.cpp
#include <iostream>
#include <vector>


int sortAndCountInversions(std::vector<int>& x, const int indexStart, const int indexEnd);
int mergeAndCountInversions(std::vector<int>& x, const int indexStart, const int indexMiddle, const int indexEnd);
void printVector(std::vector<int>& x);


int main(){
    
    //  2-1, 5-3, 5-1, 5-4, 3-1, 6-1, 6-4   ==>  7 inversions
    std::vector<int> x = {2, 5, 3, 6, 1, 4};
    
    int numInversions = sortAndCountInversions(x, 0, x.size() - 1);
    printVector(x);
    std::cout << numInversions << std::endl;
    
    return 0;
}



int sortAndCountInversions(std::vector<int>& x, const int indexStart, const int indexEnd){
    
    int count = 0;
    
    if(indexStart < indexEnd){
        
        const int indexMiddle = indexStart + (indexEnd - indexStart) / 2;
        
        count += sortAndCountInversions(x, indexStart, indexMiddle);
        count += sortAndCountInversions(x, indexMiddle + 1, indexEnd);
        count += mergeAndCountInversions(x, indexStart, indexMiddle, indexEnd);
    }
    
    return count;
}


int mergeAndCountInversions(std::vector<int>& x, const int indexStart, const int indexMiddle, const int indexEnd){
    
    int n = indexEnd - indexStart + 1;
    int* temp = new int[n];
    
    int indexTemp = 0;
    int indexLeft = indexStart;
    int indexRight = indexMiddle + 1;
    
    int count = 0;
    
    while(indexLeft <= indexMiddle && indexRight <= indexEnd){
        
        if(x[indexLeft] <= x[indexRight]){
            temp[indexTemp] = x[indexLeft];
            indexLeft++;
        }
        else{
            temp[indexTemp] = x[indexRight];
            indexRight++;
            
            int numLeftRemaining = indexMiddle - indexLeft + 1;     // number of inversions between Left and Right
            count += numLeftRemaining;
        }
        
        indexTemp++;
    }
    
    while(indexLeft <= indexMiddle){
        temp[indexTemp] = x[indexLeft];
        indexLeft++;
        indexTemp++;
    }
    
    while(indexRight <= indexEnd){
        temp[indexTemp] = x[indexRight];
        indexRight++;
        indexTemp++;
    }
    
    
    for(int i = 0; i < n; i++)
        x[indexStart + i] = temp[i];
    
    
    delete[] temp;
    
    return count;
}


void printVector(std::vector<int>& x){
    
    int n = x.size();
    
    for(int i = 0; i < n; i++)
        std::cout << x[i] << " ";
    
    std::cout << std::endl;
}

File: count-inversions/count-inversions.py
"""

Count number of inversions in an array.

e.g. [1, 3, 2, 5, 4] has inversions 3-2, 5-4

"""


def sort_count_inversions(x, index_start, index_end):
    if index_start < index_end:
        index_middle = index_start + (index_end - index_start) / 2
        count_left = sort_count_inversions(x, index_start, index_middle)
        count_right = sort_count_inversions(x, index_middle + 1, index_end)
        count_cross = merge_count_inversions(x, index_start, index_middle,
                                             index_end)
        return count_left + count_right + count_cross
    else:
        return 0


def merge_count_inversions(x, index_start, index_middle, index_end):
    n = index_end - index_start + 1
    temp = [0] * n

    index_temp = 0
    index_left = index_start
    index_right = index_middle + 1

    count = 0

    while index_left <= index_middle and index_right <= index_end:

        if x[index_left] <= x[index_right]:
            temp[index_temp] = x[index_left]
            index_left += 1
        else:
            temp[index_temp] = x[index_right]
            index_right += 1
            count += index_middle - index_left + 1

        index_temp += 1

    while index_left <= index_middle:
        temp[index_temp] = x[index_left]
        index_left += 1

    while index_right <= index_end:
        temp[index_temp] = x[index_right]
        index_right += 1

    for i in range(n):
        x[index_start + i] = temp[i]

    return count


def main():
    x = [2, 5, 3, 6, 1, 4]
    print x
    print 'Num Inversions: ' + str(sort_count_inversions(x, 0, len(x) - 1))


if __name__ == '__main__':
    main()
